File: apps/shared/src/confab-notice.ts
/**
 * Reader-side gate for the out-of-band `hermes_confab_notice` record.
 *
 * The Python half (`agent/confab_notice.py`) validates the provider payload on
 * the wire and stamps the surviving assistant row with
 * `display_kind = 'confab_notice'` plus `display_metadata.confab_notice`.
 * Every client that renders reloaded history has to answer the same question
 * — "did this turn really carry a confirmed catch?" — so the answer lives here
 * once instead of being re-derived (and re-broken) per surface.
 *
 * `display_kind` alone is NOT sufficient. It is an open-ended string column
 * that an importer, a migration, or a malformed record can populate on any
 * row. Presenting the confirmed-confabulation claim off that string would put
 * a false accusation on a user or system turn. So a row must clear two gates:
 *
 * 1. `role === 'assistant'` — only a model reply can carry a catch;
 * 2. `display_metadata.confab_notice` re-validates against the same
 *    fail-closed v1 schema the wire payload had to pass.
 *
 * Mirrors `notice_from_display_row` in `agent/confab_notice.py`.
 */

/** `display_kind` stamped on an assistant row carrying a notice. */
export const CONFAB_NOTICE_DISPLAY_KIND = 'confab_notice'

/** Key under which the validated notice sits inside `display_metadata`. */
export const CONFAB_NOTICE_KEY = 'confab_notice'

/** The only schema version this consumer understands. */
export const CONFAB_NOTICE_VERSION = 1

/** The only catch kind defined by v1 of the contract. */
export const CONFAB_NOTICE_KIND = 'scaffold_confab_removed'

/** Allowed `scope` values. */
export const CONFAB_NOTICE_SCOPES = ['visible', 'intermediate', 'both'] as const

/** Event-line label. Out of band, but never invisible. */
export const CONFAB_NOTICE_EVENT_TEXT = 'confabulation caught — scaffold text removed'

// Defensive bound — `request_id` and `grammar` are short opaque labels.
const MAX_LABEL_LEN = 256

export interface ConfabNotice {
  grammar: null | string
  kind: string
  request_id: string
  scope: string
  version: number
}

/** A history row as any of the surfaces model it. */
export interface ConfabNoticeRow {
  display_kind?: string
  display_metadata?: unknown
  role?: string
}

function isShortLabel(value: unknown): value is string {
  return typeof value === 'string' && value.trim().length > 0 && value.length <= MAX_LABEL_LEN
}

/**
 * Validate a raw notice payload. Returns the notice, or `null` on ANY
 * deviation from the v1 contract — same fail-closed posture as the producer
 * gate in Python. Never throws.
 */
export function validateConfabNotice(raw: unknown): ConfabNotice | null {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    return null
  }

  const candidate = raw as Record<string, unknown>

  if (candidate.version !== CONFAB_NOTICE_VERSION) {
    return null
  }

  if (candidate.kind !== CONFAB_NOTICE_KIND) {
    return null
  }

  if (!isShortLabel(candidate.request_id)) {
    return null
  }

  if (typeof candidate.scope !== 'string' || !CONFAB_NOTICE_SCOPES.includes(candidate.scope as never)) {
    return null
  }

  // `grammar` is the detector's bounded label, or null when several catches
  // cannot be represented by one label. Absent is treated as null.
  const grammar = candidate.grammar

  if (grammar !== undefined && grammar !== null && !isShortLabel(grammar)) {
    return null
  }

  return {
    grammar: grammar === undefined || grammar === null ? null : grammar,
    kind: CONFAB_NOTICE_KIND,
    request_id: candidate.request_id,
    scope: candidate.scope,
    version: CONFAB_NOTICE_VERSION
  }
}

/**
 * Return the validated notice a persisted row genuinely carries, else `null`.
 *
 * A remote backend older than the client can serve `display_metadata` as raw
 * JSON text, so parse a string form before reading into it.
 */
export function confabNoticeFromRow(row: ConfabNoticeRow | null | undefined): ConfabNotice | null {
  if (!row || row.role !== 'assistant' || row.display_kind !== CONFAB_NOTICE_DISPLAY_KIND) {
    return null
  }

  let metadata: unknown = row.display_metadata

  if (typeof metadata === 'string') {
    try {
      metadata = JSON.parse(metadata)
    } catch {
      return null
    }
  }

  if (!metadata || typeof metadata !== 'object') {
    return null
  }

  return validateConfabNotice((metadata as Record<string, unknown>)[CONFAB_NOTICE_KEY])
}

File: agent/confab_notice.py
"""Validation for the out-of-band ``hermes_confab_notice`` response extension.

A bridge that detects and strips self-confabulated scaffold text from a model
reply must still tell the user the reply was contaminated — that signal is
load-bearing for triage. Delivering it *in band* (appended to assistant
``content``) mutates conversation history and is replayed upstream on every
full-history recovery, so the notice moves to a top-level response-envelope
extension instead:

.. code-block:: json

    {"hermes_confab_notice": {"version": 1,
                              "kind": "scaffold_confab_removed",
                              "request_id": "3b264082",
                              "scope": "visible",
                              "grammar": "inbound"}}

This module owns the *only* gate between that untrusted provider-supplied
payload and anything user-visible or persisted. It fails closed: an unknown
version, a wrong ``kind``, a bad ``scope``, a non-string ``request_id``, or a
duplicate notice in one response yields ``None`` and a debug log. Callers must
never display or persist a payload this function rejected.

The returned object is a fresh dict containing only the validated keys — a
provider cannot smuggle extra fields into ``display_metadata`` by attaching
them to the extension.
"""

from __future__ import annotations

import logging
from typing import Any, Dict, Optional

logger = logging.getLogger(__name__)

#: Top-level key the bridge attaches to the completion / final usage chunk.
CONFAB_NOTICE_FIELD = "hermes_confab_notice"

#: Key under which the validated notice is carried on
#: ``NormalizedResponse.provider_data`` and inside ``display_metadata``.
CONFAB_NOTICE_KEY = "confab_notice"

#: ``display_kind`` stamped on the assistant row carrying a notice.
CONFAB_NOTICE_DISPLAY_KIND = "confab_notice"

#: The only schema version this consumer understands.
CONFAB_NOTICE_VERSION = 1

#: Original scaffold kind; retained for callers that distinguish status-only notices.
CONFAB_NOTICE_KIND = "scaffold_confab_removed"

# Tool guards may be the entire visible turn. Unlike the scaffold status,
# these local instructions become assistant content and may be replayed so
# the model can recover. Never interpolate provider-supplied labels here.
TOOL_CALL_NOTICE_TEXT = {
    "tool_call_unparseable": (
        "Tool call not executed: the tool-call JSON could not be parsed. "
        "Re-issue the tool call with valid JSON matching the tool schema."
    ),
    "tool_call_as_text": (
        "Tool call not executed: it was written as text rather than a native tool call. "
        "Re-issue the call using the native tool-calling interface, not a text block."
    ),
}
CONFAB_NOTICE_KINDS = (CONFAB_NOTICE_KIND, *TOOL_CALL_NOTICE_TEXT)

#: Allowed ``scope`` values.
CONFAB_NOTICE_SCOPES = ("visible", "intermediate", "both")

#: User-facing status line. Out of band, but never invisible — this is the
#: current-turn half of the triage contract.
CONFAB_NOTICE_TEXT = (
    "⚠️ Confabulation caught: the provider detected and removed "
    "self-fabricated scaffold text from this reply."
)

# Defensive bound — ``request_id`` and ``grammar`` are short opaque labels.
_MAX_LABEL_LEN = 256


def validate_confab_notice(raw: Any) -> Optional[Dict[str, Any]]:
    """Return a sanitized copy of *raw* if it is a valid v1 notice, else ``None``.

    Fails closed on every deviation from the contract. Never raises.
    """
    if not isinstance(raw, dict):
        if raw is not None:
            logger.debug(
                "Ignoring %s: expected object, got %s",
                CONFAB_NOTICE_FIELD,
                type(raw).__name__,
            )
        return None

    version = raw.get("version")
    # bool is an int subclass — True would otherwise pass as version 1.
    if isinstance(version, bool) or version != CONFAB_NOTICE_VERSION:
        logger.debug("Ignoring %s: unsupported version %r", CONFAB_NOTICE_FIELD, version)
        return None

    kind = raw.get("kind")
    if kind not in CONFAB_NOTICE_KINDS:
        logger.debug("Ignoring %s: unknown kind %r", CONFAB_NOTICE_FIELD, kind)
        return None

    request_id = raw.get("request_id")
    if not isinstance(request_id, str) or not request_id.strip():
        logger.debug(
            "Ignoring %s: request_id must be a non-empty string, got %r",
            CONFAB_NOTICE_FIELD,
            request_id,
        )
        return None
    if len(request_id) > _MAX_LABEL_LEN:
        logger.debug("Ignoring %s: request_id too long", CONFAB_NOTICE_FIELD)
        return None

    scope = raw.get("scope")
    if scope not in CONFAB_NOTICE_SCOPES:
        logger.debug("Ignoring %s: invalid scope %r", CONFAB_NOTICE_FIELD, scope)
        return None
    if kind != CONFAB_NOTICE_KIND and scope != "visible":
        logger.debug("Ignoring %s: tool-call scope must be visible", CONFAB_NOTICE_FIELD)
        return None

    # ``grammar`` is the detector's bounded label, or null when several catches
    # cannot be represented by one label. Absent is treated as null.
    grammar = raw.get("grammar")
    if grammar is not None:
        if not isinstance(grammar, str) or not grammar.strip():
            logger.debug("Ignoring %s: invalid grammar %r", CONFAB_NOTICE_FIELD, grammar)
            return None
        if len(grammar) > _MAX_LABEL_LEN:
            logger.debug("Ignoring %s: grammar too long", CONFAB_NOTICE_FIELD)
            return None

    return {
        "version": CONFAB_NOTICE_VERSION,
        "kind": kind,
        "request_id": request_id,
        "scope": scope,
        "grammar": grammar,
    }


def extract_confab_notice(obj: Any) -> Optional[Dict[str, Any]]:
    """Pull ``hermes_confab_notice`` off a completion / chunk and validate it.

    Reads the attribute first (SimpleNamespace stubs, permissive SDK models)
    and falls back to the OpenAI SDK's ``model_extra`` bag, which is where an
    unknown top-level field lands on a pydantic response model. Returns the
    sanitized notice or ``None``.
    """
    if obj is None:
        return None
    raw = getattr(obj, CONFAB_NOTICE_FIELD, None)
    if raw is None:
        extra = getattr(obj, "model_extra", None)
        if isinstance(extra, dict):
            raw = extra.get(CONFAB_NOTICE_FIELD)
    if raw is None and isinstance(obj, dict):
        raw = obj.get(CONFAB_NOTICE_FIELD)
    if raw is None:
        return None
    return validate_confab_notice(raw)
